### PR TITLE
add "requestId" to event

### PR DESCRIPTION
**Pull request by drortirosh**
requestID is a unique identifier of current request. it is a hash of the
entire UserOperation content (except the signature)

### CODE REVIEW

**Review by yoavw — COMMENTED** [review]
LGTM.  But since it's also a change in the two interfaces (IPaymaster,IWallet), it should also be changed in the EIP.

File: contracts/UserOperation.sol [hunk 1]
```diff
@@ -82,9 +82,8 @@ library UserOperationLib {
         );
     }
 
-    function hash(UserOperation calldata userOp) internal view returns (bytes32) {
-        return keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32",
-            keccak256(abi.encodePacked(pack(userOp), block.chainid))));
+    function hash(UserOperation calldata userOp) internal pure returns (bytes32) {
+        return keccak256(pack(userOp));
```
Comment (yoavw) [inline contracts/UserOperation.sol:86]: Why not include chainid and entrypoint in the request id?  Is the op-only hash useful for anything?

File: contracts/samples/SimpleWallet.sol [hunk 2]
```diff
@@ -92,25 +95,9 @@ contract SimpleWallet is IWallet {
...
-        assembly {
-            r := calldataload(signature.offset)
-            s := calldataload(add(signature.offset, 0x20))
-            v := byte(0, calldataload(add(signature.offset, 0x40)))
-        }
-    }
-
-    function _ecrecover(bytes32 hash, uint8 v, bytes32 r, bytes32 s) internal pure returns (address) {
-        return ecrecover(hash, v, r, s);
+        bytes32 k = keccak256(abi.encode(requestId, entryPoint, block.chainid));
```
Comment (yoavw) [inline contracts/samples/SimpleWallet.sol:98]: So we're still hashing twice (not saving any gas).  Just running the two hashes in separate functions.  Even if we're not flattening it to a single message (hashed just once), why not do it once, in the `userOp.hash()`?

Comment (drortirosh) [reply contracts/samples/SimpleWallet.sol]: 1. I'm not trying to optimize the last 30 gas. I'm trying to reduce high 100s ...
2. I think its cleaner this way
3. interestingly, it is also cheaper: it saves a copy of a large bytes array (yes, at the expense of an extra hash)

I added those encodings, to check the combination of pack/unpack and hash/nohash
(the "pack" method returns "bytes", and "hash" function just returns keccak(pack()))
```
    function getRequestId(UserOperation calldata userOp) public view returns (bytes32) {
        return keccak256(abi.encode(userOp.hash(), address(this), block.chainid));
    }

    function getRequestId1(UserOperation calldata userOp) public view returns (bytes32) {
        return keccak256(abi.encodePacked(userOp.hash(), address(this), block.chainid));
    }

    function getRequestId2(UserOperation calldata userOp) public view returns (bytes32) {
        return keccak256(abi.encode(userOp.pack(), address(this), block.chainid));
    }
    function getRequestId3(UserOperation calldata userOp) public view returns (bytes32) {
        return keccak256(abi.encodePacked(userOp.pack(), address(this), block.chainid));
    }
```
results: (of estimateGas in hardhat test)
0= BigNumber 26157
1= BigNumber 26152
2= BigNumber 27426
3= BigNumber 27382
that is, encodePacked is cheaper (very slightly on normal hash, little more on bytes) - but hash of hash is much cheaper than hash of bytes.
the reason, I think, is that it requires an extra copy of all the bytes in order to append the 2 words at the end, and creating array[3] is far faster.

Comment (yoavw) [reply contracts/samples/SimpleWallet.sol]: Interesting (and unintuitive to me :)
So let's go with the cheapest one.

**Pull request merged**